internal/controllers/composition/controller.go

```diff
@@ -0,0 +1,123 @@
+package composition
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	apiv1 "github.com/Azure/eno/api/v1"
+	"github.com/Azure/eno/internal/manager"
+)
+
+type compositionController struct {
+	client client.Client
+}
+
+func NewController(mgr ctrl.Manager) error {
+	c := &compositionController{
+		client: mgr.GetClient(),
+	}
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&apiv1.Composition{}).
+		WithLogConstructor(manager.NewLogConstructor(mgr, "compositionController")).
+		Complete(c)
+}
+
+func (c *compositionController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := logr.FromContextOrDiscard(ctx)
+	comp := &apiv1.Composition{}
+	err := c.client.Get(ctx, req.NamespacedName, comp)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(fmt.Errorf("getting composition resource: %w", err))
+	}
+
+	logger = logger.WithValues("compositionName", comp.Name, "compositionNamespace", comp.Namespace, "compositionGeneration", comp.Generation, "synthesisID", comp.Status.GetCurrentSynthesisUUID())
+
+	if comp.DeletionTimestamp != nil {
+		return c.reconcileDeletedComposition(ctx, comp)
+	}
+
+	if controllerutil.AddFinalizer(comp, "eno.azure.io/cleanup") {
+		err = c.client.Update(ctx, comp)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("updating composition: %w", err)
+		}
+		logger.V(1).Info("added cleanup finalizer to composition")
+		return ctrl.Result{}, nil
+	}
+
+	synth := &apiv1.Synthesizer{}
+	synth.Name = comp.Spec.Synthesizer.Name
+	err = c.client.Get(ctx, client.ObjectKeyFromObject(synth), synth)
+	if err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(fmt.Errorf("getting synthesizer: %w", err))
+	}
+	if synth != nil {
+		logger = logger.WithValues("synthesizerName", synth.Name, "synthesizerGeneration", synth.Generation)
+	}
+
+	// Enforce the synthesis timeout period
+	if syn := comp.Status.InFlightSynthesis; syn != nil && syn.Initialized != nil && synth.Spec.PodTimeout != nil {
+		delta := time.Until(syn.Initialized.Time.Add(synth.Spec.PodTimeout.Duration))
+		if delta > 0 {
+			return ctrl.Result{RequeueAfter: delta}, nil
+		}
+		syn.Canceled = ptr.To(metav1.Now())
+		if err := c.client.Status().Update(ctx, comp); err != nil {
```

Comment:
Worth patching here to avoid conflict? I'm not sure if you're relying conflict errors for anything here though

Reply:
I don't think it's a common enough occurrence to matter. But mostly I'm just afraid of coming up with the wrong `test` ops and making more weird concurrency bugs